### PR TITLE
Add new "raw" argument fix of #2052

### LINKED ISSUE #2052: Add raw variable for HTTP Web Interface
**Is your feature request related to a problem? Please describe.**
MTA WEB HTTP add variable `raw` for accepting data

**Describe alternatives you've considered**
Form-data but it is bad for example for JSONRPC

**Additional context**
https://wiki.multitheftauto.com/wiki/Resource_Web_Access
We have some variables, so we need add new:
```
    table requestHeaders: This is a table containing all the headers that were requested with the page. You can set returned headers using httpSetResponseHeader.
    table form: This is a table containing all the form data submitted to the page using HTTP POST combined with any variables passed in the querystring with HTTP GET.
    table cookies: This is a table of all the cookies. You can modify cookies using httpSetResponseCookie.
    string hostname: This is a string containing the IP address or hostname that requested the page.
    string url: This is the URL of the page.
    account user: This is the account of the current user.
```
image

### PR DESCRIPTION
**Pull request by Disinterpreter**
Fixes #2052
New "raw" argument for manual handling requests

### CODE REVIEW

**Comment by Citizen01** [comment]
hmmm what about `body` instead of `raw` ?
I don't think raw is meaningful enough as to what it is (could be the raw http request that contains the headers, cookies, body ...)
I like that WebServer is getting more features like providing the body content. The other thing I really wished it had at some point was to allow other HTTP methods (not just GET and POST) like DELETE and PUT (at least) and why not PATCH, HEAD, OPTIONS ? (I know it's out of the scope, I just wanted to let it out :p)

**Comment by Disinterpreter** [comment]
> 
> 
> hmmm what about `body` instead of `raw` ?
> I don't think raw is meaningful enough as to what it is (could be the raw http request that contains the headers, cookies, body ...)
> I like that WebServer is getting more features like providing the body content. The other thing I really wished it had at some point was to allow other HTTP methods (not just GET and POST) like DELETE and PUT (at least) and why not PATCH, HEAD, OPTIONS ? (I know it's out of the scope, I just wanted to let it out :p)

I can rename it, if other people agree with you. There's no problem.
DELETE/PUT/ETC... It does with ~raw~body. I get it. But as a problem we know nothing about type of request.

**Comment by Disinterpreter** [comment]
I implemented new argument known as **method** it return the name of type of request, like GET/POST/PUT...

you can try it with
```
<*
iprint(requestHeaders)
iprint(method)
iprint(raw)
httpSetResponseHeader ( "content-type", "application/json")
httpWrite("ok")
*>
```

File: Server/mods/deathmatch/logic/CResourceHTMLItem.cpp [hunk 1]
```diff
@@ -75,6 +75,47 @@ ResponseCode CResourceHTMLItem::Request(HttpRequest* ipoHttpRequest, HttpRespons
...
+                break;
+            case REQUESTMETHOD_UNKNOWN:
+                sMethod = "UNKNOWN";
+                break;
+            case REQUESTMETHOD_INVALID:
+                sMethod = "INVALID";
+                break;
+            default:
+                sMethod = "INVALID";
+                break;
```
Comment (Citizen01) [inline Server/mods/deathmatch/logic/CResourceHTMLItem.cpp]:
```suggestion
            case REQUESTMETHOD_INVALID:
            default:
                sMethod = "INVALID";
                break;
```

**Review by Citizen01 — CHANGES_REQUESTED** [review]
And I still want it to be named `body`. Here are my arguments:

1. You are exposing the `sBody` attribute of the HttpRequest why not naming it `body` ?
2. `sMethod` is exposed as `method`
3. `headers` list is exposed as `headers`
4. `formData` as `form`, `cookies` as `cookies`, then why `sBody` as `raw` ?
5. I think the `raw` only came from postman ?

**Comment by Citizen01** [comment]
I don't think there is a poll system so let's do it like this:

If you want the body to be exposed as `raw`,  react with ❤️ 
If you want the body to be exposed as `body`, react with 🚀

File: Server/mods/deathmatch/logic/CResourceHTMLItem.cpp [hunk 2]
```diff
@@ -75,6 +75,45 @@ ResponseCode CResourceHTMLItem::Request(HttpRequest* ipoHttpRequest, HttpRespons
             headers.PushString((*iter).second.c_str());
         }
 
+        std::string sMethod;
```
Comment (botder) [inline Server/mods/deathmatch/logic/CResourceHTMLItem.cpp]:
```suggestion
        const char* sMethod = "INVALID";
```
Is the heap allocation inside `std::string` really necessary, if you only assign strings from static storage?
With the suggestion you can also simplify the `default` case blow.
`CLuaArguments::PushString` should support pointer to const char.

**Comment by botder** [comment]
> If you want the body to be exposed as `body`, react with 🚀

What about `requestBody`? That would be analog to `requestHeaders`.

**Comment by Disinterpreter** [comment]
> 
> 
> > If you want the body to be exposed as `body`, react with 🚀
> 
> What about `requestBody`? That would be analog to `requestHeaders`.

sounds fine for me, and I think it same for @Citizen01 so I did it. And added code suggestions of @botder

**Review by botder — APPROVED** [review]

**Review by Citizen01 — APPROVED** [review]
Tried to brake it sending weird postman requests but everthing I tried worked as expected so it's okay for me too.
Thanks for this PR, this will help me getting back into the API project I had and will simplify the whole thing too ! 👍 🚀

**Pull request merged**